### PR TITLE
allow multiple signatures

### LINKED ISSUE #14: Consolidate signature algorithm definitions
There are currently 2 different algorithm definitions (one for NSS and one for COSE). Let's make this only one.

### LINKED ISSUE #15: Make sign and verify work with the actual format we use
The COSE signature that can be processed at the moment are not exactly the ones we want. They should look like this

    98(
     [
       / protected / h'..', / List of DER encoded intermediate certificates /
       / unprotected / {},
       / payload / nil, / The payload is the contents of the manifest file /
       / signatures / [
         [
           / protected / h'a2012604..' / {
               \ alg \ 1:-7, \ ECDSA with SHA-256 \
               \ kid \ 4:'..' \ DER encoded signing certificate \
             } / ,
           / unprotected / {},
           / signature / h'e2ae..'
         ],
         [
           / protected / h'a201382404..' / {
               \ alg \ 1:-37, \ RSASSA-PSS with SHA-256 \
               \ kid \ 4:'..' \ DER encoded signing certificate \
             } / ,
           / unprotected / {},
           / signature / h'00a2..'
         ]
       ]
     ]
   )

Changes necessary are:
- [x]  Move `kid` to the protected signature header
- [x] Add intermediate certificates to the protected body header
- [x] Put the EE certificate in `kid` instead of the SPKI
- [x] Make the payload `nil`
- [ ] Handle multiple signatures

Work is happening  [here](https://github.com/franziskuskiefer/cose-rust/tree/correct-signature-format).

### PR DESCRIPTION
**Pull request by franziskuskiefer**
This will fix #15.
It's based on PR #34 so shouldn't be merged. I'll make the PR against master when #34 got merged.
This also fixes #14.

### CODE REVIEW

File: src/cose/cose_sign.rs [hunk 1]
```diff
@@ -1,37 +1,40 @@
...
+    ee_certs: &[&[u8]],
     cert_chain: &[&[u8]],
-    pkcs8: &[u8],
+    pkcs8: &[&[u8]],
 ) -> Result<Vec<u8>, CoseError> {
-    let nss_alg = match alg {
-        CoseSignatureType::ES256 => nss::SignatureAlgorithm::ES256,
-        CoseSignatureType::PS256 => nss::SignatureAlgorithm::PS256,
-    };
+    assert_eq!(pkcs8.len(), ee_certs.len());
```
Comment (mozkeeler) [inline src/cose/cose_sign.rs]: Looks like we should define a type that represents the trio (algorithm, certificate, key).

File: src/cose/mod.rs [hunk 2]
```diff
@@ -34,32 +34,43 @@ pub enum CoseError {
...
-        signature_bytes,
-    );
-    if !verify_result.is_ok() {
-        return Err(CoseError::VerificationFailed);
+    for signature in cose_signatures {
+        let signature_algorithm = &signature.signature_type;
+        let signature_bytes = &signature.signature;
+        let real_payload = &signature.to_verify;
+
+        println!("Verifying signature {:?}", signature_algorithm);
```
Comment (mozkeeler) [inline src/cose/mod.rs]: nit: probably don't need debug output in the checked-in version

File: src/cose/util.rs [hunk 3]
```diff
@@ -94,42 +110,34 @@ pub fn get_sig_struct_bytes(
 //  ]
 pub fn build_cose_signature(
     cert_chain: &[&[u8]],
-    ee_cert: &[u8],
-    sig_bytes: &[u8],
-    alg: &CoseSignatureType,
+    ee_certs: &[&[u8]],
+    sig_bytes: &Vec<Vec<u8>>,
+    alg: &[SignatureAlgorithm],
```
Comment (mozkeeler) [inline src/cose/util.rs]: Similar idea here - I think we should have a vec of a new type that encapsulates these values.

File: src/cose/decoder.rs [hunk 4]
```diff
@@ -144,19 +102,68 @@ pub fn decode_signature(bytes: Vec<u8>, payload: &[u8]) -> Result<Vec<CoseSignat
...
         protected_signature_header_serialized.clone(),
         payload,
     );
 
-    let signature = CoseSignature {
+    Ok(CoseSignature {
         signature_type: signature_algorithm,
         signature: signature_bytes,
         signer_cert: ee_cert,
         certs: Vec::new(),
```
Comment (mozkeeler) [inline src/cose/decoder.rs]: Do we want to grab these out of the protected_body_head here?

File: src/cose/decoder.rs [hunk 5]
```diff
@@ -129,8 +87,8 @@ pub fn decode_signature(bytes: Vec<u8>, payload: &[u8]) -> Result<Vec<CoseSignat
     let signature_algorithm = match signature_algorithm {
         CborType::SignedInteger(val) => {
             match val {
-                -7 => CoseSignatureType::ES256,
-                -37 => CoseSignatureType::PS256,
+                -7 => SignatureAlgorithm::ES256,
+                -37 => SignatureAlgorithm::PS256,
```
Comment (mozkeeler) [inline src/cose/decoder.rs]: I was going to say we should add -35 / ES384 here, but I guess we don't actually support decoding it yet, so let's wait until we have test data for it.

Comment (franziskuskiefer) [reply src/cose/decoder.rs]: yeah, we need some more tests for ES384 (and ES512 if we want to support that).

File: src/cose/decoder.rs (hunk 4 above)
Comment (franziskuskiefer) [reply src/cose/decoder.rs]: We should :D That's why I want to move on #10 so that we actually use all the parts and don't mess up things like this.

**Comment by franziskuskiefer** [comment]
Closing this, meging #38.

**Pull request closed without merging**